### DISCUSSION TITLE
The distant future of the EVM

### DISCUSSION THREAD

**Discussion by alexauroradev** [near/nearcore#3719] [EVM]
This discussion has arisen from the comment: https://github.com/near/nearcore/issues/3699#issuecomment-741473118

The question that we need to make a decision about: what is the future of the EVM Runtime? Should it stay as Wasm runtime in the nearcore forever, or it should be moved out somehow at some point?

**Comment by alexauroradev** [comment]
@ilblackdragon, @artob, @nearmax, @evgenykuzyakov, @bowenwang1996, @SkidanovAlex your inputs would be valuable here.

**Reply by bowenwang1996** [reply]
Forever is a strong term and that is definitely not decided. But I think short to medium term, it seems reasonable to have evm stay in nearcore.

**Reply by ilblackdragon** [reply]
At this point it will stay in nearcore for sure.
Just let's not make any decisions that prevent EVM's removal as precompile back into a contract later.
As far as I know, currently that is the case.

I do think that long term we will have more user-defined runtimes (e.g. have their own account models), and it make sense to have highly optimized compiled WASM for those and for other frequently used contracts.

E.g. Specific set of LLVM compiled contracts (similar to what Polkadot is doing to their parachain runtimes) that allows only 2-3x performance hit, which given most of the performance issues are in storage probably is enough. 

Such "trusted" contracts can be decided by technical governance when it's clear that there is no JIT bombs in such contracts.

**Reply by artob** [reply]
I understand the appeal of the original approach of having the EVM be a contract instead of a precompile. From a risk management perspective, though, it seems somewhat unlikely that we would endeavor to move it back to being a contract once the EVM is deployed and in active use by numerous customers. The risk/reward calculation there doesn't seem to have any good payoff that I'm aware of. There are other concerns as well, but that's for starters.

**Comment by alexauroradev** [comment]
Final decision: the EVM is going to be in the contract level (at least glue code), with optimised compiling and several parts of it going in the protocol level, like execution of the EVM code. The implementation of the EVM contract can be found in https://github.com/near/near-evm/
